apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/BufferedFile.java

```diff
@@ -0,0 +1,244 @@
+package co.elastic.apm.agent.profiler.asyncprofiler;
+
+import co.elastic.apm.agent.objectpool.Recyclable;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * An abstraction similar to {@link MappedByteBuffer} that allows to read the content of a file with an API that is similar to
+ * {@link ByteBuffer}.
+ * <p>
+ * Instances of this class hold a reusable buffer that contains a subset of the file,
+ * or the whole file if the buffer's capacity is greater or equal to the file's size.
+ * </p>
+ * <p>
+ * Whenever calling a method like {@link #getLong()} or {@link #position(long)} would exceed the currently buffered range
+ * the same buffer is filled with a different range of the file.
+ * </p>
+ * <p>
+ * The downside of {@link MappedByteBuffer} (and the reason for implementing this abstraction)
+ * is that calling methods like {@link MappedByteBuffer#get()} can increase time-to-safepoint.
+ * This is because these methods are implemented as JVM intrinsics.
+ * When the JVM executes an intrinsic, it does not switch to the native execution context which means that it's not ready to enter a safepoint
+ * whenever a intrinsic runs.
+ * As reading a file from disk can get stuck (for example when the disk is busy) calling {@link MappedByteBuffer#get()} may take a while to execute.
+ * While it's executing other threads have to wait for it to finish if the JVM wants to reach a safe point.
+ * </p>
+ */
+class BufferedFile implements Recyclable {
+
+    private static final int SIZE_OF_BYTE = 1;
+    private static final int SIZE_OF_SHORT = 2;
+    private static final int SIZE_OF_INT = 4;
+    private static final int SIZE_OF_LONG = 8;
+    private final ByteBuffer buffer;
+    private final int capacity;
+    private long offset;
+    private long limit;
+    private boolean wholeFileInBuffer;
+    @Nullable
+    private RandomAccessFile randomAccessFile;
+    @Nullable
+    private FileChannel fileChannel;
+
+    public BufferedFile(ByteBuffer buffer) {
+        this.buffer = buffer;
+        capacity = buffer.capacity();
+    }
+
+    /**
+     * Starts reading the file into the {@linkplain #buffer buffer}
+     *
+     * @param file the file to read from
+     * @throws IOException If some I/O error occurs
+     */
+    public void setFile(File file) throws IOException {
+        randomAccessFile = new RandomAccessFile(file, "r");
+        fileChannel = randomAccessFile.getChannel();
+        read(0);
```

Comment:
Seems like we read redundant 4MB at the beginning, since after getting metadata offset, we immediately seek to the end of the file. Probably, `read` can accept another parameter - the maximum number of bytes to read.

Reply:
good catch! done.